File: hydro/src/main/java/com/ultreon/hydro/screen/ScreenManager.java
package com.ultreon.hydro.screen;

import com.ultreon.commons.annotation.FieldsAreNonnullByDefault;
import com.ultreon.commons.annotation.MethodsReturnNonnullByDefault;
import com.ultreon.hydro.Game;
import com.ultreon.hydro.common.ResourceEntry;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nullable;

import javax.annotation.ParametersAreNonnullByDefault;

/**
 * Scene manager, used for change between scenes.
 *
 * @see Screen
 */
@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
@FieldsAreNonnullByDefault
public class ScreenManager {
    private static ScreenManager instance;
    private static final Logger LOGGER = LogManager.getLogger("Screen-Manager");
    private final Screen startScreen;
    private final Game game;

    @Nullable
    private Screen currentScreen;
    private boolean initialized = false;

    public static ScreenManager getInstance() {
        return instance;
    }

    private ScreenManager(Screen startScreen) {
        instance = this;
        this.game = Game.getInstance();
        this.currentScreen = this.startScreen = startScreen;
    }

    public static ScreenManager create(Screen start) {
        return new ScreenManager(start);
    }

    /**
     * Display a new scene.
     *
     * @param scene the scene to display
     * @return if changing the scene was successful.
     */
    @SuppressWarnings("UnusedReturnValue")
    public boolean displayScreen(@Nullable Screen scene) {
        return displayScreen(scene, false);
    }

    /**
     * Display a new scene.
     *
     * @param screen the scene to display
     * @return if changing the scene was successful.
     */
    public boolean displayScreen(@Nullable Screen screen, boolean force) {
        if (currentScreen != null) {
            LOGGER.debug("Hiding " + currentScreen.getClass().getSimpleName());
            if (currentScreen.onClose(screen) || force) {
                return initNextScreen(screen);
            }
            LOGGER.debug("Hiding " + currentScreen.getClass().getSimpleName() + " canceled.");
        } else {
            LOGGER.debug("Hiding <<NO-SCENE>>");
            return initNextScreen(screen);
        }
        return false;
    }

    private boolean initNextScreen(@Nullable Screen screen) {
        this.currentScreen = screen;
        if (screen != null) {
            game.getGameWindow().setCursor(screen.getDefaultCursor());
        } else {
            game.getGameWindow().setCursor(game.getDefaultCursor());
        }

        if (currentScreen != null) {
            LOGGER.debug("Showing " + currentScreen.getClass().getSimpleName());
            this.currentScreen.init();
        } else {
            LOGGER.debug("Showing <<NO-SCENE>>");
        }
        return true;
    }

    /**
     * DEPRECATED
     *
     * @return the current scene resource location.
     * @deprecated since 1.0.???
     */
    @Contract("->null")
    @Deprecated
    @Nullable
    public ResourceEntry getCurrentSceneKey() {
        return null;
    }

    public synchronized void start() {
        if (initialized) {
            throw new IllegalStateException("SceneManager already initialized.");
        }

        this.initialized = true;
        this.currentScreen = this.startScreen;
        LOGGER.debug("Showing " + currentScreen.getClass().getSimpleName());
        game.getGameWindow().setCursor(this.startScreen.getDefaultCursor());
        this.startScreen.init();
    }

    @Nullable
    public Screen getCurrentScreen() {
        return currentScreen;
    }

    public boolean isInitialized() {
        return initialized;
    }
}

File: src/main/kotlin/qtech/bubbles/common/scene/ScreenManager.kt
package qtech.bubbles.common.scene

import qtech.bubbles.BubbleBlaster
import qtech.bubbles.annotation.FieldsAreNonnullByDefault
import qtech.bubbles.annotation.MethodsReturnNonnullByDefault
import qtech.bubbles.common.ResourceLocation
import qtech.bubbles.common.screen.Screen
import qtech.bubbles.screen.LoadScreen
import qtech.bubbles.util.Util
import org.apache.logging.log4j.LogManager
import org.jetbrains.annotations.Contract
import javax.annotation.ParametersAreNonnullByDefault

/**
 * Scene manager, used for change between scenes.
 *
 * @see Screen
 */
@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
@FieldsAreNonnullByDefault
class ScreenManager private constructor(private val startScreen: Screen) {
    var currentScreen: Screen?
        private set
    var isInitialized = false
        private set

    private val onStart: MutableList<() -> Unit> = ArrayList()

    /**
     * Display a new scene.
     *
     * @param scene the scene to display
     * @return if changing the scene was successful.
     */
    fun displayScreen(scene: Screen?): Boolean {
        return displayScreen(scene, false)
    }

    /**
     * Display a new scene.
     *
     * @param screen the scene to display
     * @return if changing the scene was successful.
     */
    fun displayScreen(screen: Screen?, force: Boolean): Boolean {
        if (currentScreen != null) {
            LOGGER.debug("Hiding " + currentScreen!!.javaClass.simpleName)
            if (currentScreen!!.onClose(screen) || force) {
                currentScreen = screen
                if (screen != null) {
                    Util.setCursor(screen.defaultCursor)
                } else {
                    Util.setCursor(BubbleBlaster.instance.defaultCursor)
                }
                if (currentScreen != null) {
                    LOGGER.debug("Showing " + currentScreen!!.javaClass.simpleName)
                    currentScreen!!.init()
                } else {
                    LOGGER.debug("Showing <<NO-SCENE>>")
                }
                return true
            }
            LOGGER.debug("Hiding " + currentScreen!!.javaClass.simpleName + " canceled.")
        } else {
            LOGGER.debug("Hiding <<NO-SCENE>>")
            currentScreen = screen
            if (screen != null) {
                Util.setCursor(screen.defaultCursor)
            } else {
                Util.setCursor(BubbleBlaster.instance.defaultCursor)
            }
            if (currentScreen != null) {
                LOGGER.debug("Showing " + currentScreen!!.javaClass.simpleName)
                currentScreen!!.init()
            } else {
                LOGGER.debug("Showing <<NO-SCENE>>")
            }
            return true
        }
        return false
    }

    /**
     * DEPRECATED
     *
     * @return the current scene resource location.
     */
    @get:Deprecated("since 1.0.???", ReplaceWith("null"))
    @get:Contract("->null")
    @Deprecated("since 1.0.???", ReplaceWith("null"))
    val currentSceneKey: ResourceLocation?
        get() = null

    @Synchronized
    fun start() {
        check(!this.isInitialized) { "SceneManager already initialized." }
        this.isInitialized = true
        currentScreen = startScreen
        LOGGER.debug("Showing " + currentScreen!!.javaClass.simpleName)
        Util.setCursor(startScreen.defaultCursor)
        startScreen.init()
    }

    fun onStart(function: () -> Unit) {
        onStart.add(function)
    }

    companion object {
        @JvmStatic
        val instance = ScreenManager(LoadScreen())
        private val LOGGER = LogManager.getLogger("QB:Screen:Manager")
    }

    init {
        currentScreen = startScreen
    }
}